src/js/common/validator.js

```diff
@@ -1,122 +1,540 @@
+     */
+    gpii.schema.validator.validate = function (toValidate, gssSchema, ajvOptions) {
+        ajvOptions = ajvOptions || gpii.schema.validator.defaultAjvOptions;
+
+        // Validate the GSS schema against the metaschema before proceeding.
+        var schemaValidationResults = gpii.schema.validator.validateSchema(gssSchema, ajvOptions);
+        if (schemaValidationResults.isError) {
+            return schemaValidationResults;
         }
-    }
+        else {
+            var ajv = new Ajv(ajvOptions);
+            ajv.addMetaSchema(gpii.schema.metaSchema);
 
-    return undefined;
-};
-
-
-/**
- *
- * Iterate through the raw validation errors and evolve them if possible.  This is gated and will abort if the parser is
- * not yet ready.  This function is called once the parser is ready.
- *
- * @param that - The validator component itself.
- * @param schemaKey {String} - The filename/id of the schema we are working with.
- * @param rawErrors {Object} - The raw error data returned by AJV.
- * @return An {Object} representing the original error data combined with any "evolved" error messages we were able to find.
- */
-gpii.schema.validator.ajv.sanitizeValidationErrors = function (that, schemaKey, rawErrors) {
-    var schemaContent = that.model.schemas[schemaKey];
-    var evolvedErrors = fluid.transform(rawErrors, function (error) {
-        return gpii.schema.errors.evolveError(schemaContent, error);
-    });
-
-    return evolvedErrors;
-};
-
-/**
- *
- * If we receive new schemas, make the validator aware of them so that we can simply validate using their key.
- *
- * @param that - The validator component itself.
- *
- */
-gpii.schema.validator.ajv.refreshSchemas = function (that) {
-    // Update the list of schemas using the supplied content
-    fluid.each(that.model.schemas, function (schemaContent, schemaKey) {
-        // AJV will not let us overwrite an existing schema , so we have to remove the current content first.
-        if (that.ajv.getSchema(schemaKey)) {
-            that.ajv.removeSchema(schemaKey);
+            // We have to validate against a transformed copy of the original rawSchema so that AJV can enforce our
+            // required fields, which it would otherwise ignore.
+            var rawSchema = gpii.schema.gssToJsonSchema(gssSchema);
+            var validator = ajv.compile(rawSchema);
+            validator(toValidate);
+
+            return gpii.schema.validator.standardiseAjvErrors(gssSchema, validator.errors);
         }
+    };
+
+    /**
+     *
+     * Validate a "GPII Schema System" schema.
+     *
+     * @param {GssSchema} gssSchema - A GSS schema definition.
+     * @param {Object} ajvOptions - Optional arguments to pass to the underlying AJV validator.
+     * @return {Object} - An object that describes the results of validation.  The `isValid` property will be `true` if the data is valid, or `false` otherwise.  The `isError` property will be set to `true` if there are validation errors.
```

Comment:
Presumably this should use the newly created @typdef

Reply:
Sorry, ignore earlier comment - I see that this is an entirely different structured return, which nonetheless should have a @typedef of its own

Reply:
Done.